<launch>
    <arg name="world_frame" default="world" />
    <arg name="robot_density" default="100" />
    <arg name="robot_height" default="0.5" />
    <arg name="robot_width" default="0.46" />
    <arg name="robot_frame" default="P1" />
    <arg name="cmd_vel_topic" default="/RosAria/cmd_vel" />
    <arg name="scan_topic" default="/RosAria/scan" />

    <node pkg="tf2_ros" type="static_transform_publisher" name="base_link_to_laser_tf_publisher"
    args="0.1 0 0 0 0 -0.98480775 -0.17364818 P1 laser" /> <!-- -20 -180deg in yaw -->

    <node pkg="tf2_ros" type="static_transform_publisher" name="map_to_world_tf_publisher" 
    args="-1.31144779 1.18 0.29090403 -0.000115597 0.003501224 -0.000026968 0.999993864 world map" />

    <!-- <node pkg="tf2_ros" type="static_transform_publisher" name="Pioneer_to_base_link_tf_publisher" 
    args="0 0 0 0 0 0 1 base_link P1" /> -->

    <node name="map_server" pkg="map_server" type="map_server" args="$(find rrt_project)/maps/real_map2/real_map.yaml" />

    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find rrt_project)/rviz/real_world.rviz" />

    <node name="teleop" pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py" output="screen" >
        <remap from="cmd_vel" to="$(arg cmd_vel_topic)"/>
    </node>

    <node pkg="rpic_article" type="emergency_button.py" name="emergency_button" output="screen" required="true"/>
    <!-- <include file="$(find vrpn_client_ros)/launch/sample.launch"/> -->
    <arg name="new_robot_pose_topic" default="/vrpn_client_node/P1/pose" />
    <arg name="new_robot_pose_message" default="PoseStamped" />
    <arg name="new_robot_control_topic" default="/RosAria/cmd_vel" />
    <arg name="new_robot_control_message" default="Twist" />
    <arg name="x_goal" default="3" />
    <arg name="y_goal" default="0" />

    <group ns="P1">
        <param name="scan_topic" value="$(arg scan_topic)"/>
        <param name="robot_frame" value="$(arg robot_frame)"/>
        <param name="world_frame" value="$(arg world_frame)"/>
        <param name="robot_density" value="$(arg robot_density)"/>
        <param name="robot_height" value="$(arg robot_height)"/>
        <param name="robot_width" value="$(arg robot_width)"/>

        <node name="P1_velocity_calculation" pkg="rrt_project" type="velocity_publisher" output="screen">
            <param name="robot_frame" value="$(arg robot_frame)" />
            <param name="world_frame" value="$(arg world_frame)" />
            <param name="rate" value="50" />
            <param name="velocity_topic" value="velocity" />
        </node>

        <node pkg="rrt_project" type="differential_robot_route.py" name="pioneer_node" output="screen">
            <rosparam file="$(find rrt_project)/config/differential_robots/pioneer.yaml" command="load" />
            <param name="robot_pose_topic" value="$(arg new_robot_pose_topic)" />
            <param name="robot_pose_message" value="$(arg new_robot_pose_message)" />
            <param name="robot_control_topic" value="$(arg new_robot_control_topic)" />
            <param name="robot_control_message" value="$(arg new_robot_control_message)" />
        </node>

        <node pkg="rrt_project" type="rrt_path_node" name="rrt_path_node" output="screen" >
            <rosparam file="$(find rrt_project)/config/rrt.yaml" command="load" />
            <param name="robot_frame" value="$(arg robot_frame)" />
            <param name="x_goal" value="$(arg x_goal)" />
            <param name="y_goal" value="$(arg y_goal)" />
        </node>

        <node pkg="rrt_project" type="publish_robot_potential_points" name="publish_robot_points" output="screen" >
            <param name="world_frame" value="$(arg world_frame)" />
            <param name="robot_frame" value="$(arg robot_frame)" />
            <param name="robot_density" value="$(arg robot_density)" />
            <param name="robot_height" value="$(arg robot_height)" />
            <param name="robot_width" value="$(arg robot_width)" />
        </node>

        <node pkg="rrt_project" type="obstacle_avoidance_node.py" name="obstacle_avoidance_node" output="screen">
            <rosparam file="$(find rrt_project)/config/obstacle_avoidance/obstacle_avoidance_scan_node.yaml" command="load" />
            <param name="scan_topic" value="$(arg scan_topic)"/>
            <param name="robot_frame" value="$(arg robot_frame)"/>
            <param name="world_frame" value="$(arg world_frame)"/>
            <param name="observation_radius_gain" value="1.5"/>
            <param name="min_observation_radius" value="0.25"/>
            <param name="velocity_topic" value="velocity"/>
        </node>
    </group>
</launch>